lib/evaluate/shape.cc

```diff
@@ -349,8 +353,28 @@ Shape GetUpperBounds(FoldingContext &context, const NamedEntity &base) {
 auto GetShapeHelper::operator()(const Symbol &symbol) const -> Result {
   return std::visit(
       common::visitors{
-          [&](const semantics::ObjectEntityDetails &) {
-            return (*this)(NamedEntity{symbol});
+          [&](const semantics::ObjectEntityDetails &object) {
+            if (IsImpliedShape(symbol)) {
+              return (*this)(object.init());
+            } else {
+              Shape shape;
+              int n{object.shape().Rank()};
+              NamedEntity base{symbol};
+              for (int dimension{0}; dimension < n; ++dimension) {
+                shape.emplace_back(GetExtent(context_, base, dimension));
+              }
+              return Result{shape};
+            }
+          },
+          [&](const semantics::EntityDetails &) {
```

Comment:
The reference capture specification is not needed here or below for the default case.